File: src/Web/JavaScriptJS/patterns/observer.js
const assert = require("node:assert/strict");
/** Publishes state changes to independently subscribed observers. */
function run() {
  const listeners = new Set();
  const seen = [];
  const subject = {
    subscribe(listener) { listeners.add(listener); return () => listeners.delete(listener); },
    publish(value) { for (const listener of listeners) listener(value); },
  };
  const unsubscribe = subject.subscribe((value) => seen.push(value));
  subject.publish("ready");
  unsubscribe();
  subject.publish("ignored");
  assert.deepEqual(seen, ["ready"]);
  return "observer";
}
module.exports = { run };
if (require.main === module) run();

File: src/Scripting/Lua/patterns/observer.lua
local subject = { subscribers = {} }
function subject:subscribe(fn) table.insert(self.subscribers, fn) end
function subject:publish(value) for _, fn in ipairs(self.subscribers) do fn(value) end end
local seen = {}
subject:subscribe(function(v) table.insert(seen, "a" .. v) end)
subject:subscribe(function(v) table.insert(seen, "b" .. v) end)
subject:publish("1")
assert(table.concat(seen, ",") == "a1,b1")
return true
